asf_heat_pump_suitability/pipeline/run_scripts/calculate_garden_size.py

```diff
@@ -0,0 +1,140 @@
+"""
+Script to calculate garden area (m2) where possible for properties in the domestic EPC register.
+"""
+
+import argparse
+import logging
+import pandas as pd
+from tqdm import tqdm
+import polars as pl
+import geopandas as gpd
+from argparse import ArgumentParser
+from asf_heat_pump_suitability.pipeline.prepare_features import (
+    lat_lon,
+    land_extent,
+    building_footprint,
+    garden_size,
+)
+
+
+def argparser() -> argparse.Namespace:
+    """
+    Create ArgumentParser and parse.
+
+    Returns:
+        argparse.Namespace: object holding argument attributes
+    """
+    parser = ArgumentParser()
+
+    parser.add_argument(
+        "--use_mapping",
+        help="Path to existing mapping of land extent files to LAD boundary geometries",
+        type=str,
+        required=False,
+    )
+
+    parser.add_argument(
+        "--epc_path",
+        help="Path to EPC file with properties to estimate garden size for",
+        type=str,
+        required=False,
+    )
+
+    parser.add_argument(
+        "--save_epc_gardens",
+        help="Path to save output file with garden size per EPC record to",
+        type=str,
+        required=False,
+    )
+
+    parser.add_argument(
+        "--save_land_file_bounds",
+        help="Path to save land extent file bounds to",
+        type=str,
+        required=False,
+    )
+
+    args = parser.parse_args()
+
+    return args
+
+
+if __name__ == "__main__":
+    _args = argparser()
+
+    # Load EPC x, y coordinates in CRS: EPSG:27700
+    epc_gdf = pl.read_parquet(
+        _args.epc_path, columns=["UPRN", "X_COORDINATE", "Y_COORDINATE"]
+    )
+    epc_gdf = lat_lon.generate_gdf_uprn_coords(epc_gdf)[["UPRN", "geometry"]]
+
+    if not _args.use_mapping:
+        # Get land extent file boundaries
+        land_file_bounds = land_extent.generate_gdf_map_file_to_bounds(
+            save_as=_args.save_land_file_bounds
+        )
+    else:
+        # Load existing file with land extent files mapped to LAD boundaries
+        land_file_bounds = gpd.read_file(_args.use_mapping, crs="EPSG:27700")
+
+    # Get building footprint file boundaries
+    microsoft_file_bounds = building_footprint.transform_df_uk_dataset_links()
+
+    # Check where building footprint files and land extent files overlap
+    file_matches = garden_size.match_dict_files_land_building(
+        land_files_gdf=land_file_bounds, building_files_gdf=microsoft_file_bounds
+    )
+
+    epc_gardens = []
+    prev = None
+    total_gardens = 0
+    for land_file, building_file in tqdm(file_matches.items()):
```

Comment:
This loop could probably be abstracted to a pipeline function rather than being part of the script.

Reply:
The reason I left this in the script is because when we run it in interactive mode with `python -i`, we can more easily address errors if they come up without having to re-run the whole pipeline (as it took 24hr+ before the suggested speed improvements were implemented). I'll keep as is for now until we're sure it's robust and/or the pipeline is significantly sped up.